spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java

```
@@ -0,0 +1,125 @@
+
+    return taskCommitRDD.collect().stream()
+        .flatMap(taskCommit -> Arrays.stream(taskCommit.files()))
+        .collect(Collectors.toList());
+  }
+
+  private Writer.TaskCommit rewriteDataForTask(CombinedScanTask task) {
+    TaskContext context = TaskContext.get();
+
+    RowDataReader dataReader = new RowDataReader(task, SchemaParser.fromJson(tableSchema),
+        SchemaParser.fromJson(tableSchema), fileIO.value(), encryptionManager.value(), caseSensitive);
+
+    int partitionId = context.partitionId();
+    long taskId = context.taskAttemptId();
+    DataWriter<InternalRow> dataWriter = writerFactory.createDataWriter(partitionId, taskId, 0);
+
+    Throwable originalThrowable = null;
+    try {
+      while (dataReader.next()) {
+        InternalRow row = dataReader.get();
+        dataWriter.write(row);
+      }
+
+      dataReader.close();
+      dataReader = null;
+      return (Writer.TaskCommit) dataWriter.commit();
+
+    } catch (Throwable t) {
+      originalThrowable = t;
+      try {
+        LOG.error("Aborting task", originalThrowable);
+        context.markTaskFailed(originalThrowable);
+
+        LOG.error("Aborting commit for partition {} (task {}, attempt {}, stage {}.{})",
+            partitionId, taskId, context.attemptNumber(), context.stageId(), context.stageAttemptNumber());
+        if (dataReader != null) {
+          dataReader.close();
+        }
+        dataWriter.abort();
+        LOG.error("Aborted commit for partition {} (task {}, attempt {}, stage {}.{})",
+            partitionId, taskId, context.taskAttemptId(), context.stageId(), context.stageAttemptNumber());
+
+      } catch (Throwable inner) {
+        if (originalThrowable != inner) {
+          originalThrowable.addSuppressed(inner);
+          LOG.warn("Suppressing exception in catch: {}", inner.getMessage(), inner);
+        }
+      }
+
+      throw new RuntimeException(originalThrowable);
```

Instead of wrapping by default, this should check if the throwable is an `Exception` and throw it, then wrap if it isn't an exception.